tools/automatic_query_fixer/src/main/java/com/google/cloud/bigquery/utils/queryfixer/fixer/TableNotFoundFixer.java

```diff
@@ -0,0 +1,122 @@
+    this.bigQueryService = bigQueryService;
+    this.queryPositionConverter = new QueryPositionConverter(query);
+  }
+
+  @Override
+  public FixResult fix() {
+    TableId fullTableId = constructTableId(err.getTableName());
+    List<String> tableNames =
+        bigQueryService.listTableNames(fullTableId.getProject(), fullTableId.getDataset());
+
+    StringUtil.SimilarStrings similarTables =
+        StringUtil.findSimilarWords(tableNames, fullTableId.getTable());
+
+    // this is an arbitrary standard. It requires the candidate table should share at least 50%
+    // similarity as
+    // the incorrect table typo.
+    int editDistanceThreshold = (fullTableId.getTable().length() + 1) / 2;
+
+    if (similarTables.getStrings().isEmpty()
+        || similarTables.getDistance() > editDistanceThreshold) {
+      return FixResult.failure(err);
+    }
+
+    // this method only finds the first occurrence of the incorrect table. It is possible that this
+    // table exists in multiple positions of this query. What is worse, it is possible that the
+    // table name is also part of a literal, then this auto fixing may have problem. The ultimate
+    // solution for this issue is to use Parser to find the correct position of this table.
+    int tableStartIndex = findTheIndexOfIncorrectTable();
+
+    List<FixOption> fixOptions =
+        similarTables.getStrings().stream()
+            .map(
+                table -> {
+                  String fullTableName =
+                      constructFullTableName(
+                          fullTableId.getProject(), fullTableId.getDataset(), table);
+                  String fixedQuery = replaceTable(fullTableName, tableStartIndex);
+                  return FixOption.of(fullTableName, fixedQuery);
+                })
+            .collect(Collectors.toList());
+
+    return FixResult.success(/*approach= */ "Replace the table name.", fixOptions, err);
+  }
+
+  private TableId constructTableId(String fullTableName) {
+    List<String> contents = PatternMatcher.extract(fullTableName, tableIdRegex);
+    String projectId;
+    // Assume the table name from dry-run message is always correct, so no check is performed for
+    // matching.
+    // If contents[0] == null, it means the table looks like dataset.table, so the project ID should
+    // be fetched from the BigQuery Client.
+    if (contents.get(0) == null) {
+      projectId = bigQueryService.getBigQueryOptions().getProjectId();
+    } else {
+      projectId = contents.get(1);
+    }
+    String datasetId = contents.get(2);
+    String tableName = contents.get(3);
+    return TableId.of(projectId, datasetId, tableName);
+  }
+
+  private String constructFullTableName(String projectId, String datasetId, String tableName) {
+    return String.format("%s.%s.%s", projectId, datasetId, tableName);
+  }
+
+  private int findTheIndexOfIncorrectTable() {
+    // the table in the error message is presented in the legacySQL mode, but this fixer is used to
```

Comment:
So we rely on the error message to identify if they used legacy sql?

Reply:
Here I mean the error message represents the table in legacySQL format, which is `project:dataset.table`. I guess we could be an issue for BigQuery to update this error message.